client/src/components/assessments/OndemandAssessments.js

```diff
@@ -0,0 +1,340 @@
+import { Formik } from "formik"
+import moment from "moment"
+import { PeriodsDetailsPropType } from "periodUtils"
+import PropTypes from "prop-types"
+import React, { useCallback, useEffect, useMemo, useState } from "react"
+import { Button, Panel, Table } from "react-bootstrap"
+import { toast } from "react-toastify"
+import REMOVE_ICON from "resources/delete.png"
+import Settings from "settings"
+
+const GQL_DELETE_NOTE = gql`
+  mutation($uuid: String!) {
+    deleteNote(uuid: $uuid)
+  }
+`
+
+const OnDemandAssessments = ({
+  entity,
+  entityType,
+  style,
+  periodsDetails,
+  canAddAssessment,
+  onUpdateAssessment
+}) => {
+  /* recurrence has the value 'ondemand' for this specific assessment type and
+      numberOfPeriods is a property of the parent component (AssessmentResultsContainer)
+      which is used to determine how many columns should be displayed inside of the
+      ondemand assessment table. Recalculated according to the screensize. */
+  const { recurrence, numberOfPeriods } = periodsDetails
+  // Button text.
+  const addAssessmentLabel = `Make a new ${entity?.toString()} assessment`
+  const [showModal, setShowModal] = useState(false)
+  // Used to determine if the AssessmentModal is in edit mode or create mode.
+  const [editModeObject, setEditModeObject] = useState({
+    questionareResults: {},
+    uuid: ""
+  })
+  // 'assessmentConfig' has question set for ondemand assessments defined in the dictionary
+  // and 'assessmentYupSchema' used for this question set.
+  const {
+    assessmentConfig,
+    assessmentYupSchema
+  } = entity.getPeriodicAssessmentDetails(recurrence)
+
+  const filteredAssessmentConfig = Model.filterAssessmentConfig(
+    assessmentConfig,
+    entity
+  )
+
+  // Cards array updated before loading the page & after every save of ondemand assessment.
+  const assessmentCards = useMemo(() => {
+    const cards = []
+    const sortedOnDemandNotes = entity.getOndemandAssessments()
+    sortedOnDemandNotes.forEach((note, index) => {
+      const parentFieldName = `assessment-${note.uuid}`
+      const assessmentFieldsObject = JSON.parse(note.text)
+      cards.push(
+        <>
+          <div
+            style={
+              sortedOnDemandNotes.length - 1 !== index
+                ? {
+                  color: "red",
+                  paddingBottom: "3px",
+                  margin: "0 -1rem 1rem 0",
+                  borderBottom: "2px solid lightgrey"
+                }
+                : {
+                  paddingBottom: "3px",
+                  marginBottom: "1rem",
+                  borderBottom: "2px solid lightgrey"
+                }
+            }
+          >
+            <b>
+              {/* Only the last object in the sortedOnDemandNotes can be valid.
+                  If the expiration date of the last object is older than NOW,
+                  it is also expired. */}
+              {sortedOnDemandNotes.length - 1 !== index
```

Comment:
I think `index !== sortedOnDemandNotes.length - 1` makes more sense here as index is changing through each iteration and rhs stays constant.